substrate/frame/delegated-staking/src/types.rs

```diff
@@ -0,0 +1,308 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Basic types used in delegated staking.
+
+use super::*;
+use frame_support::traits::DefensiveSaturating;
+
+/// The type of pot account being created.
+#[derive(Encode, Decode)]
+pub(crate) enum AccountType {
+	/// A proxy delegator account created for a nominator who migrated to an `Agent` account.
+	///
+	/// Funds for unmigrated `delegator` accounts of the `Agent` are kept here.
+	ProxyDelegator,
+}
+
+/// Information about delegation of a `delegator`.
+#[derive(Default, Encode, Clone, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct Delegation<T: Config> {
+	/// The target of delegation.
+	pub agent: T::AccountId,
+	/// The amount delegated.
+	pub amount: BalanceOf<T>,
+}
+
+impl<T: Config> Delegation<T> {
+	/// Get delegation of a `delegator`.
+	pub(crate) fn get(delegator: &T::AccountId) -> Option<Self> {
+		<Delegators<T>>::get(delegator)
+	}
+
+	/// Create and return a new delegation instance.
+	pub(crate) fn from(agent: &T::AccountId, amount: BalanceOf<T>) -> Self {
+		Delegation { agent: agent.clone(), amount }
+	}
+
+	/// Ensure the delegator is either a new delegator or they are adding more delegation to the
+	/// existing agent.
+	///
+	/// Delegators are prevented from delegating to multiple agents at the same time.
+	pub(crate) fn can_delegate(delegator: &T::AccountId, agent: &T::AccountId) -> bool {
+		Delegation::<T>::get(delegator)
+			.map(|delegation| delegation.agent == agent.clone())
+			.unwrap_or(
+				// all good if it is a new delegator except it should not be an existing agent.
+				!<Agents<T>>::contains_key(delegator),
+			)
+	}
+
+	/// Checked decrease of delegation amount. Consumes self and returns a new copy.
+	pub(crate) fn decrease_delegation(self, amount: BalanceOf<T>) -> Option<Self> {
+		let updated_delegation = self.amount.checked_sub(&amount)?;
+		Some(Delegation::from(&self.agent, updated_delegation))
+	}
+
+	/// Checked increase of delegation amount. Consumes self and returns a new copy.
+	#[allow(unused)]
+	pub(crate) fn increase_delegation(self, amount: BalanceOf<T>) -> Option<Self> {
+		let updated_delegation = self.amount.checked_add(&amount)?;
+		Some(Delegation::from(&self.agent, updated_delegation))
+	}
+
+	/// Save self to storage. If the delegation amount is zero, remove the delegation.
+	pub(crate) fn save_or_kill(self, key: &T::AccountId) {
```

Comment:
nit

```suggestion
	pub(crate) fn update_or_kill(self, key: &T::AccountId) {
```

Reply:
This is good, and same as `fn put(self)` I imagined: you are done with this, it should not remain in scope, therefore consume self, and write back to storage. If it is zero, writing is basically deleting, else updating.